src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedEntryPointSymbol.cs

```diff
@@ -329,6 +330,169 @@ private static BoundCall CreateParameterlessCall(CSharpSyntaxNode syntax, BoundE
+                }
+                else
+                {
+                    return compilation.GetSpecialType(SpecialType.System_Void);
+                }
+            }
+
+            private static ReturnKind GetReturnKind(CSharpCompilation compilation, TypeSymbol returnType) {
+                if (returnType.IsGenericTaskType(compilation)) {
+                    return ReturnKind.IntReturning;
+                }
+                else
+                {
+                    return ReturnKind.VoidReturning;
+                }
+            }
+
+            internal AsyncForwardEntryPoint(CSharpCompilation compilation, DiagnosticBag diagnosticBag, NamedTypeSymbol containingType, MethodSymbol userMain): 
+                base(containingType, TranslateReturnType(compilation, userMain.ReturnType)) {
+                Debug.Assert(userMain != null);
+
+                _userMain = userMain;
+                _diagnosticBag = diagnosticBag;
+                _compilation = compilation;
+                _paramKind = userMain.ParameterCount == 0 ? ParamKind.NoArgs : ParamKind.StringArrayArgs;
+                _returnKind = GetReturnKind(compilation, userMain.ReturnType);
+
+                switch (_paramKind)
+                {
+                    case ParamKind.StringArrayArgs:
+                        var stringType = _compilation.GetSpecialType(SpecialType.System_String);
+                        var stringArrayType = ArrayTypeSymbol.CreateCSharpArray(_compilation.Assembly, stringType);
+                        _parameters = ImmutableArray.Create(
+                            SynthesizedParameterSymbol.Create(this, stringArrayType, 0, RefKind.None));
+                        break;
+                    default:
+                        _parameters = ImmutableArray<ParameterSymbol>.Empty;
+                        break;
+                }
+            }
+
+            public override string Name => MainName;
+
+            public override ImmutableArray<ParameterSymbol> Parameters => _parameters;
+
+            internal override BoundBlock CreateBody()
+            {
+                var syntax = this.GetSyntax();
+
+                var refKinds = Parameters.Length == 0  ? ImmutableArray<RefKind>.Empty : ImmutableArray.Create(RefKind.None);
+                var arguments = new List<BoundExpression>();
+                foreach (var parameterSymbol in Parameters)
+                {
+                    arguments.Add(new BoundParameter(syntax, parameterSymbol, parameterSymbol.Type));
+                }
+
+                // Main(args) or Main()
+                BoundCall userMainInvocation = new BoundCall(
+                        syntax: this.GetSyntax(),
+                        receiverOpt: null,
+                        method: _userMain,
+                        arguments: arguments.ToImmutableArray(),
+                        argumentNamesOpt: ImmutableArray<string>.Empty,
```

Comment:
I think that you meant for this comment to be on another line though.  Our main method parameter names is defined in the call to SynthesizedParameterSymbol.Create inside the constructor.